XLS-0066d-lending-protocol/README.md

```diff
@@ -0,0 +1,1428 @@
+
+        // A late payment increases the value of the loan by the difference between periodic and late payment interest
+        return (late_payment.principal, late_payment.interest, late_payment.interest - periodic_payment.interest, loan.late_payment_fee)
+    }
+
+    let full_payment = loan.compute_full_payment(current_time)
+
+    // if the payment is equal or higher than full payment amount
+    // and there is more than one payment remaining, make a full payment
+    if amount >= full_payment && loan.payments_remaining > 1 {
+        loan.payments_remaining = 0
+        loan.principal_outstanding = 0
+
+        // A full payment decreases the value of the loan by the difference between the interest paid and the expected outstanding interest
+        return (full_payment.principal, full_payment.interest, full_payment.interest - loan.compute_current_value().interest, full_payment.fee)
+    }
+
+    // if the payment is not late nor if it's a full payment, then it must be a periodic once
+
+    let periodic_payment = loan.compute_periodic_payment()
+
+    let full_periodic_payments = floor(amount / periodic_payment)
+    if full_periodic_payments < 1 {
+        return "insufficient amount paid" error
+    }
+
+    loan.payments_remaining -= full_periodic_payments
+    loan.next_payment_due_date = loan.next_payment_due_date + loan.payment_interval * full_periodic_payments
+    loan.last_payment_date = loan.next_payment_due_date - loan.payment_interval
+
+
+    let total_principal_paid = 0
+    let total_interest_paid = 0
+    let loan_value_change = 0
+    let total_fee_paid = loan.service_fee * full_periodic_payments
+
+    while full_periodic_payments > 0 {
+        total_principal_paid += periodic_payment.principal
+        total_interest_paid += periodic_payment.interest
+        periodic_payment = loan.compute_periodic_payment()
+        full_periodic_payments -= 1
+    }
+
+    loan.principal_outstanding -= total_principal_paid
+
+    let overpayment = min(loan.principal_outstanding, amount % periodic_payment)
+    if overpayment > 0 && is_set(lsfOverayment) {
+        let interest_portion = overpayment * loan.overpayment_interest_rate
+        let fee_portion = overpayment * loan.overpayment_fee
+        let remainder = overpayment - interest_portion - fee_portion
+
+        total_principal_paid += remainder
+        total_interest_paid += interest_portion
+        total_fee_paid += fee_portion
+
+        let current_value = loan.compute_current_value()
+        loan.principal_outstanding -= remainder
+        let new_value = loan.compute_current_value()
+
+        loan_value_change = (new_value.interest - current_value.interest) + interest_portion
+    }
+
+    return (total_principal_paid, total_interest_paid, loan_value_change, total_fee_paid)
+```
+
+##### 3.2.5.3 Failure Conditions
+
+- A `Loan` object with specified `LoanID` does not exist on the ledger.
+
+- The Loan has not started yet: `Loan.StartDate > CurrentTime`.
```

Comment:
If the loan allows early payments, why is this an impediment?

Reply:
In essence this allows to pre-create a loan for some future time.

Reply:
> In essence this allows to pre-create a loan for some future time.

Ok, so it's like the loan doesn't exist until the `StartDate`, so it doesn't make sense to make a payment, because you're making a payment on nothing.